src/reservoir/Reservoir.ts

```diff
@@ -0,0 +1,226 @@
+import {
+  createClient,
+  getClient,
+  // @ts-ignore
+  // eslint-disable-next-line import/no-unresolved
+} from "@reservoir0x/reservoir-sdk";
+import {
+  Address,
+  createPublicClient,
+  getContract,
+  Hash,
+  http,
+  PublicClient,
+} from "viem";
+import { mainnet } from "viem/chains";
+
+import { Wallet, zeroHash } from "@/blockchain";
+import { erc721ABI } from "@/generated/blockchain/v5";
+
+import {
+  adaptWalletToCaptureTxData,
+  ETH_CONTRACT_ADDRESS,
+  generateExpectedCurrencyPriceObject,
+  generateMatchOrdersExecutionData,
+  getExecutionData,
+  SeaportAskOrBid,
+  WETH_CONTRACT_ADDRESS,
+} from "./utils";
+
+export class Reservoir {
+  baseApiUrl: string;
+  mainnetClient: PublicClient;
+  wallet: Wallet;
+
+  constructor({
+    baseApiUrl = "https://api.reservoir.tools",
+    apiKey,
+    infuraApiKey,
+    wallet,
+  }: {
+    baseApiUrl?: string;
+    apiKey?: string;
+    infuraApiKey?: string;
+    wallet: Wallet;
+  }) {
+    this.baseApiUrl = baseApiUrl;
+    this.wallet = wallet;
+
+    createClient({
+      chains: [
+        {
+          id: wallet.chain.id,
+          name: wallet.chain.name,
+          baseApiUrl,
+          active: true,
+        },
+      ],
+      apiKey,
+      source: "gondi.xyz",
+    });
+
+    this.mainnetClient = createPublicClient({
+      chain: mainnet,
+      transport: http(`https://mainnet.infura.io/v3/${infuraApiKey}`),
+    });
+  }
+
+  async getAsk({ orderId }: { orderId: Hash }) {
+    return fetch(
+      `${this.baseApiUrl}/orders/asks/v5?ids=${orderId}&includeRawData=true`
+    )
+      .then((res) => res.json() as Promise<{ orders: unknown[] }>)
+      .then(({ orders }) => orders[0] as SeaportAskOrBid);
+  }
+
+  async getBid({ orderId }: { orderId: Hash }) {
+    return fetch(
+      `${this.baseApiUrl}/orders/bids/v6?ids=${orderId}&includeRawData=true`
+    )
+      .then((res) => res.json() as Promise<{ orders: unknown[] }>)
+      .then(({ orders }) => orders[0] as SeaportAskOrBid);
+  }
+
+  async getExecutionDataForBuyToken({
+    collectionContractAddress,
+    tokenId,
+    price,
+    exactOrderSource,
+  }: {
+    collectionContractAddress: string;
+    tokenId: bigint;
+    price: bigint;
+    exactOrderSource?: string;
+  }) {
+    const { adaptedWallet, txData } = adaptWalletToCaptureTxData(this.wallet);
+
+    try {
+      await getClient()?.actions.buyToken({
+        items: [
+          {
+            token: `${collectionContractAddress}:${tokenId}`,
+            quantity: 1,
+            exactOrderSource,
+          },
+        ],
+        expectedPrice: generateExpectedCurrencyPriceObject(
+          price,
+          ETH_CONTRACT_ADDRESS
+        ),
+        wallet: adaptedWallet,
+        onProgress: () => null,
+        precheck: false,
+        options: {
+          excludeEOA: true,
+          skipBalanceCheck: true,
+        },
+      });
+    } catch (e) {
```

Comment:
Here we'd catch the custom error. I know no other error can happen, but I think it's a cleaner interface and easier to read